runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/net/HostResolver.kt

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.net
+
+import aws.smithy.kotlin.runtime.util.InternalApi
+
+/**
+ * A resolved (hostname, address) pair
+ */
+public data class HostAddress(
+    /**
+     * The name that [address] was resolved from
+     */
+    val hostname: String,
+
+    /**
+     * The resolved internet address
+     */
+    val address: IpAddr,
+)
+
+/**
+ * Component capable of resolving host names to one or more internet addresses
+ */
+@InternalApi
+public interface HostResolver {
+    public companion object {
+        /**
+         * The default DNS host resolver (usually the default for the platform, e.g. InetAddress for JVM)
+         */
+        public val Default: HostResolver = DefaultHostResolver
+    }
+
+    /**
+     * Resolves the address(es) for a particular host and returns the list
+     * of addresses
+     */
+    public suspend fun resolve(hostname: String): List<HostAddress>
```

Comment:
**question**

Ref arch calls out this function taking a struct, e.g.:

```kotlin
data class HostResolverArgs(val hostname: String, val service?: String = null)

public interface HostResolver {
     public suspend fun resolve(args: HostResolverArgs): List<HostAddress>
}
```

I shied away from this because (1) there is no documented use for it yet and (2) we can always add an overload later. I have no real strong opinion though, thoughts?